presto-main/src/main/java/com/facebook/presto/sql/planner/SortExpressionExtractor.java

```diff
@@ -59,7 +63,34 @@ private SortExpressionExtractor() {}
             return Optional.empty();
         }
 
-        return new SortExpressionVisitor(buildSymbols).process(filter);
+        List<Expression> filterConjuncts = ExpressionUtils.extractConjuncts(filter);
+        SortExpressionVisitor visitor = new SortExpressionVisitor(buildSymbols);
+
+        List<SortExpressionContext> sortExpressionCandidates = filterConjuncts.stream()
+                .map(visitor::process)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(groupingBy(SortExpressionContext::getSortExpression, reducing(SortExpressionExtractor::merge)))
+                .values()
+                .stream()
+                .map(Optional::get)
+                .collect(toImmutableList());
+
+        // For now heuristically pick sort expression which has most search expression assigned to it.
+        // TODO: make it cost based decision based on symbol statistics
+        return sortExpressionCandidates
+                .stream()
+                .sorted(comparing(context -> -context.getSearchExpressions().size()))
```

Comment:
Consider using `.reversed()` instead of `-`

Reply:
I tried that before. As well as `Comparator.reverseOrder`. But then I have to add explicit cast of `context` to `SortExpressionContext`.
Like this:
```
                .sorted(comparing(context -> ((SortExpressionContext) context).getSearchExpressions().size()).reversed())
```

I prefer `-` in that case unless you can help me get rid of cast.

Reply:
Alas. However, rather than cast, you should rather `s/context ->/(SortExpressionContext context) ->`.

You can also `-1 * ` instead of `-` to make it more visible. Or leave as is, not a big deal.